### DISCUSSION TITLE
@WithSpan getting context

### DISCUSSION THREAD

**Discussion by aborigene** [open-telemetry/opentelemetry-java-instrumentation#10130] [Q&A]
Hi,

I have a situation where I have a listener code, that listens to SQS and on each message, this method is executed. The messages are produced by a Lambda function, that is adding otel parameters traceid and tracestate to message attributes.

On the Java side, I had to create TextMapGetter and retrieve the context. Is there a way of just annotation the class in a way that this context would also be retrieved?

Code snippet below:
`@SpringBootApplication
public class BankFargateApplication {

    OpenTelemetry openTelemetry = OpenTelemetrySdk.builder()
            .setPropagators(ContextPropagators.create(W3CTraceContextPropagator.getInstance()))
            .buildAndRegisterGlobal();

    Tracer tracer = openTelemetry.getTracer("dynatrace-manual-opentel", "1.0.0");

    public static void main(String[] args) {
        SpringApplication.run(BankFargateApplication.class, args);
    }

    public class StringToHexadecimal {

        String getValue(String integer) {
            StringBuffer sb = new StringBuffer();
            // Converting string to character array
            char ch[] = integer.toCharArray();
            // System.out.println("Converted: "+
            // Long.toHexString(Integer.parseInt(integer)));
            System.out.println("Converted: " + String.format("%x", integer));
            for (int i = 0; i < ch.length; i++) {
                String hexString = Integer.toHexString(ch[i]);

                System.out.println("ch[i]: " + ch[i]);
                System.out.println("hexString: " + hexString);
                sb.append(hexString);
            }
            System.out.println("Hex Conversion: " + sb.toString());
            return (sb.toString());
        }
    }

    TextMapGetter<JSONObject> getter = new TextMapGetter<>() {
        @Override
        public String get(JSONObject carrier, String key) {
            //System.out.println("TextMapGetter - " + key);

            JSONObject carrier_data = (JSONObject) carrier.get("traceparent");
            String value = carrier_data.get("Value").toString();
            JSONObject propagationValue = new JSONObject(value);
            String traceparent = propagationValue.get("traceparent").toString();

            if (traceparent != null) {
                return traceparent;
            }
            return null;
        }

        @Override
        public Iterable<String> keys(JSONObject carrier) {
            return null;
        }

    };

    public void processMessage(ReceiveMessageRequest receiveRequest, SqsClient sqsClient) throws Exception {
        List<Message> messages = sqsClient.receiveMessage(receiveRequest).messages();

        // System.out.println(messages);

        for (Message m : messages) {
            // System.out.println("\n" +m.body());

            try {
                JSONObject messageBody = new JSONObject(m.body());
                JSONObject mattrib = (JSONObject) messageBody.get("MessageAttributes");
                TextMapPropagator propagator = GlobalOpenTelemetry.getPropagators().getTextMapPropagator();
                Context parentContext = propagator.extract(Context.current(), mattrib, getter);
                Span span = tracer.spanBuilder("process message")
                        .setParent(parentContext)
                        .setSpanKind(SpanKind.SERVER)
                        .startSpan();

                System.out.println("parentContext: " + parentContext);

                try {
                    // Add the attributes defined in the Semantic Conventions
                    span.setAttribute("key", "value");

                } finally {
                    span.end();
                }

            } catch (NullPointerException e) {
                System.out.println("Encontrado valor nulo.");
            }
        }

    }`

I would like to annotage te processMessage method, to create a more simple change to application using annotations as we have tons of similar applications, could this be done in anyway?

**Comment by laurit** [comment]
> I would like to annotage te processMessage method, to create a more simple change to application using annotations as we have tons of similar applications, could this be done in anyway?

Sure it can. Sky is the limit! You could post process the bytecode or use an annotation processor to find the annotated methods and from the method with the piece of code that iterates the messages and for each message extract the context and create the span. This probably can be done but is a bit hard. You could simplify the problem by changing your requirements to instead annotating a method that takes a single message as argument. For example if you have something like `@WithMessage public void process(Message message) {...}` then you could probably just use spring aop, for example see https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/main/instrumentation/spring/spring-boot-autoconfigure/src/main/java/io/opentelemetry/instrumentation/spring/autoconfigure/instrumentation/annotations/InstrumentationWithSpanAspect.java If you further relax your requirements and drop the need for annotations you could also just have the code that extracts the context and starts the span in some util method or in a base class.
